File: 2015/cpp/day02.cc
#include <algorithm>
#include <cstdlib>
#include <numeric>
#include <string>
#include <vector>

#include "utils.h"

class Package {
public:
  Package(int length, int width, int height)
      : length_(length), width_(width), height_(height){};

  static Package make_package(const std::string &s) {
    auto parts = aoc::utils::split(s, 'x');
    std::vector<int> lengths;
    for (const auto &part : parts) {
      lengths.push_back(std::stoi(part));
    }
    if (lengths.size() != 3) {
      throw std::runtime_error("Invalid package: " + s);
    }

    return Package(lengths[0], lengths[1], lengths[2]);
  }

  int wrapping_paper() const {
    auto sides = {length_ * width_, width_ * height_, height_ * length_};
    return 2 * std::accumulate(sides.begin(), sides.end(), 0) +
           *std::min_element(sides.begin(), sides.end());
  }

  int ribbon() const {
    auto perimeters = {2 * (length_ + width_), 2 * (width_ + height_),
                       2 * (height_ + length_)};
    return *std::min_element(perimeters.begin(), perimeters.end()) +
           length_ * width_ * height_;
  }

private:
  int length_;
  int width_;
  int height_;
};

int p1(const std::vector<Package> &packages) {
  int total = 0;
  for (const auto &p : packages) {
    total += p.wrapping_paper();
  }
  return total;
}

int p2(const std::vector<Package> &packages) {
  int total = 0;
  for (const auto &p : packages) {
    total += p.ribbon();
  }
  return total;
}

int main() {
  std::vector<std::string> lines = aoc::utils::get_trimmed_lines();
  std::vector<Package> packages;

  for (const auto &line : lines) {
    try {
      packages.push_back(Package::make_package(line));
    } catch (const std::invalid_argument &exception) {
      std::cerr << "Error parsing '" << line << "': " << exception.what()
                << std::endl;
      return 1;
    }
  }

  try {
    std::cout << "Part 1: " << p1(packages) << std::endl;
    std::cout << "Part 2: " << p2(packages) << std::endl;
  } catch (const std::exception &exception) {
    std::cerr << exception.what() << std::endl;
    return 1;
  }
}

File: 2015/rust/src/bin/day02.rs
use anyhow::Result;

use aoc::utils::{get_input, parse_fields};

fn main() -> Result<()> {
    let input = get_input()?;
    let boxes: Vec<Box> = parse_fields(input.trim(), '\n')?;

    println!("Part 1: {}", part1(&boxes));
    println!("Part 2: {}", part2(&boxes));

    Ok(())
}

fn part1(boxes: &[Box]) -> u64 {
    boxes.iter().map(Box::wrapping_paper).sum()
}

fn part2(boxes: &[Box]) -> u64 {
    boxes.iter().map(Box::ribbon).sum()
}

struct Box {
    length: u64,
    width: u64,
    height: u64,
}

impl Box {
    fn wrapping_paper(&self) -> u64 {
        let sides = [
            self.length * self.width,
            self.width * self.height,
            self.height * self.length,
        ];
        let surface_area = 2 * sides.iter().sum::<u64>();

        sides.iter().min().unwrap() + surface_area
    }

    fn ribbon(&self) -> u64 {
        let perimeters = [
            2 * (self.length + self.width),
            2 * (self.width + self.height),
            2 * (self.height + self.length),
        ];
        let volume = self.length * self.width * self.height;

        perimeters.iter().min().unwrap() + volume
    }
}

mod parsing {
    use super::Box;

    use anyhow::bail;

    impl std::str::FromStr for Box {
        type Err = anyhow::Error;

        fn from_str(s: &str) -> Result<Self, Self::Err> {
            let dimensions: Vec<u64> = s
                .split('x')
                .map(|x| x.parse())
                .collect::<std::result::Result<_, _>>()?;
            if dimensions.len() != 3 {
                bail!("Failed to parse line: {}", s);
            }

            Ok(Box {
                length: dimensions[0],
                width: dimensions[1],
                height: dimensions[2],
            })
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Box;

    #[test]
    fn wrapping_paper_1() {
        let b: Box = "2x3x4".parse().unwrap();
        assert_eq!(b.wrapping_paper(), 58);
    }

    #[test]
    fn wrapping_paper_2() {
        let b: Box = "1x1x10".parse().unwrap();
        assert_eq!(b.wrapping_paper(), 43);
    }

    #[test]
    fn ribbon_1() {
        let b: Box = "2x3x4".parse().unwrap();
        assert_eq!(b.ribbon(), 34);
    }
    #[test]
    fn ribbon_2() {
        let b: Box = "1x1x10".parse().unwrap();
        assert_eq!(b.ribbon(), 14);
    }
}
